asio/CMakeLists.txt

```diff
@@ -0,0 +1,240 @@
+
+if(CONAN_COMPILER)
+
+  # ==========================================================
+  project(asio VERSION ${CONAN_PACKAGE_VERSION} LANGUAGES CXX)
+  # ==========================================================
+
+  include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+  conan_basic_setup(TARGETS)    # this set CONAN_TARGETS
+
+else()
+
+  # use ccache and clang++ if found
+  find_program(CCACHE_EXECUTABLE "ccache" HINTS /usr/local/bin /opt/local/bin)
+  if(CCACHE_EXECUTABLE)
+    message(STATUS "use ccache")
+    set(CMAKE_CXX_COMPILER_LAUNCHER "${CCACHE_EXECUTABLE}" CACHE PATH "ccache" FORCE)
+    set(CMAKE_C_COMPILER_LAUNCHER "${CCACHE_EXECUTABLE}" CACHE PATH "ccache" FORCE)
+  endif()
+
+  # ========================================
+  project(asio VERSION 1.14.1 LANGUAGES CXX)
+  # ========================================
+
+endif()
+
+#----------------------------------------------------------
+# Compiler config
+#----------------------------------------------------------
+if(NOT DEFINED CMAKE_CXX_STANDARD)
+  set(CMAKE_CXX_STANDARD 17)
+  set(CMAKE_CXX_STANDARD_REQUIRED ON)
+  set(CMAKE_CXX_EXTENSIONS OFF)
+endif()
+message(STATUS "USING: CMAKE_CXX_STANDARD=${CMAKE_CXX_STANDARD}")
+message(STATUS "USING: CMAKE_CXX_STANDARD_REQUIRED=${CMAKE_CXX_STANDARD_REQUIRED}")
+message(STATUS "USING: CMAKE_CXX_EXTENSIONS=${CMAKE_CXX_EXTENSIONS}")
+
+
+###########################################################
+option(ASIO_STANDALONE "Build asio without boost" ON)
+###########################################################
+if(ASIO_STANDALONE)
+  option(COMPILER_WARNINGS_ARE_ERRORS "To be pedantic! ;-)" ON)
+  if(COMPILER_WARNINGS_ARE_ERRORS)
+    if(MSVC)
+      # warning level 4 and all warnings as errors
+      add_compile_options(/W4 /WX)
+    else()
+      # lots of warnings and all warnings as errors
+      add_compile_options(-Wall -Wextra -Wpedantic -Werror)
+    endif()
+  endif()
+else()
+  unset(COMPILER_WARNINGS_ARE_ERRORS CACHE)
+endif()
+###########################################################
+
+
+# =========================================================
+# ASIO Options
+#==========================================================
+option(ASIO_NO_DEPRECATED "Build asio without deprecated interfaces" ON)
+option(ASIO_BUILD_TESTS "Build asio unittest too" OFF)
+option(ASIO_BUILD_EXAMPLES "Build asio examples too" OFF)
+option(ASIO_ENABLE_HANDLER_TRACKING "enable handler tracking" OFF)
+option(ASIO_DISABLE_KQUEUE "disable kqueue" OFF)
+option(ASIO_DISABLE_EPOLL "disable epoll" OFF)
+option(ASIO_USE_CLANG_TIDY "use clang-tidy if found" OFF)
+
+
+#
+# Where to put all the RUNTIME targets when built.  This variable is used
+# to initialize the RUNTIME_OUTPUT_DIRECTORY property on all the targets.
+#
+set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/bin)
+set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/lib)
+set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/lib)
+
+
+set(CPPdefinitions)
+set(EXTRA_LIBS)
```

Comment:
Everything related to these variables should just go in the respective `target_*` commands, but wrapped in generator expressions.